rfcs/20230515-DynamicEmbedding.md

```diff
@@ -0,0 +1,250 @@
+recommendation models have become more complicated, there are three distinct
+challenges that need to be addressed. These include the difficulty in
+separating popular and less-popular items or adapting to the seasonal cycle
+of popularity, the lack of a cross-platform solution for handling larger
+and larger embedding tables the dynamic nature of large embedding tables
+due to modeling large unique id-based features and the crossing features
+among them.
+
+Currently, there are two ways to handle such limitations in TensorFlow: 
+direct hashing without a vocabulary
+a pre-computed fixed vocab with out-of-vocabulary hashing.
+Neither approximation gives the user a fine grained control over
+vocab-embedding mapping.  Hence, the proposal aims to provide a native
+solution for handling these challenges by introducing the concept of
+DynamicEmbedding.
+
+### Why Keras?
+We believe that internal and external users share many common pain points.
+To support these features, external users today often need to rebuild an
+entire suite of APIs, including optimizers, distributed training logic,
+and customized TF kernels, to work around TensorFlow restrictions (that
+variables are special-cased). As the middle layer of the TF tech stack,
+we believe that we are in the best position to work with upstream 1P and
+3P users, consolidate feedback, collaborate to drive a hardware-agnostic
+solution.
+
+## User Benefit
+This initiative offers several benefits, including:
+Providing a unified TensorFlow solution that allows for productive
+exploration and potential large model quality gain across different use
+cases.
+Reducing computation cost and training latency by eliminating the need
+for a pre-computed vocab.
+Strengthening TensorFlow's advantage for third-party adoption.(Nvidia,
+spotify, Tencent/Alibaba
+(RFC: https://github.com/tensorflow/recommenders-addons/blob/master/rfcs/20200424-sparse-domain-isolation.md)
+- vip.com case study
+(https://drive.google.com/file/d/1UEWtixlA_zucLLkXlmgbF-4DAZHHKNmo/view?resourcekey=0-QXC4KOuQ6_RXuaYiyyRfYQ)
+
+Additionally, many external users that rely on TensorFlow have already
+adopted this idea, and open-source libraries have been pushing on this
+front - TorchRec with a native embedding distribution and a dynamic
+embedding solution & HugeCTR (Merlin) with a highly-performant
+embedding caching strategy. This makes it essential for TensorFlow to
+introduce a native solution to stay competitive in the market.
+
+## Design Proposal
+In this design approach, the DynamicEmbedding layer is composed of two
+layers: the DynamicLookup layer and the Embedding layer. The
+DynamicLookup layer is responsible for the following tasks:
+    * Maintaining a vocabulary table using an eviction policy that is 
+    updated based on input pattern.
+    * Performing vocabulary lookup for the given input and returning
+    integer indexes.
+    * The index is then passed to the Embedding layer, which looks
```

Comment:
In many case, we don't know how many keys in a feature exactly, since the property of videos/images/commodity/video-games are always in change. Preset a vocab/index range may lead to waste in storage or feature conflicts.